cub/cub/device/dispatch/dispatch_streaming_reduce.cuh

```diff
@@ -0,0 +1,451 @@
+
+    // Transforms the per-partition result to a global result by adding the current partition's offset to the arg result
+    // of a partition
+    promote_to_global_op_t promote_to_global_op{GlobalOffsetT{0}};
+
+    accumulating_transform_output_op_t accumulating_out_op(
+      nullptr, nullptr, false, d_result_out, promote_to_global_op, reduce_op);
+
+    // Upper bound at which we want to cut the input into multiple partitions
+    static constexpr PerPartitionOffsetT max_partition_size = ::cuda::std::numeric_limits<PerPartitionOffsetT>::max();
+
+    // Whether the given number of items fits into a single partition
+    const bool is_single_partition =
+      static_cast<GlobalOffsetT>(max_partition_size) >= static_cast<GlobalOffsetT>(num_items);
+
+    // The largest partition size ever encountered
+    const auto largest_partition_size =
+      is_single_partition ? static_cast<PerPartitionOffsetT>(num_items) : max_partition_size;
+
+    empty_problem_init_t initial_value{{PerPartitionOffsetT{1}, init}};
+
+    // Query temporary storage requirements for per-partition reduction
+    dispatch_reduce_t::Dispatch(
+      nullptr,
+      allocation_sizes[0],
+      d_indexed_offset_in,
+      THRUST_NS_QUALIFIER::make_tabulate_output_iterator(accumulating_out_op),
+      static_cast<PerPartitionOffsetT>(largest_partition_size),
+      reduce_op,
+      initial_value,
+      stream);
+
+    // Alias the temporary allocations from the single storage blob (or compute the necessary size
+    // of the blob)
+    cudaError_t error = cub::AliasTemporaries(d_temp_storage, temp_storage_bytes, allocations, allocation_sizes);
+    if (error != cudaSuccess)
+    {
+      return error;
+    }
+
+    // Return if the caller is simply requesting the size of the storage allocation
+    if (d_temp_storage == nullptr)
+    {
+      return cudaSuccess;
+    }
+
+    // Pointer to the double-buffer of global accumulators, which aggregate cross-partition results
+    global_accum_t* const d_global_aggregates = reinterpret_cast<global_accum_t*>(allocations[1]);
+
+    accumulating_out_op = accumulating_transform_output_op_t{
+      d_global_aggregates,
+      (d_global_aggregates + 1),
+      is_single_partition,
+      d_result_out,
+      promote_to_global_op,
+      reduce_op};
+
+    for (GlobalOffsetT current_partition_offset = 0; current_partition_offset < static_cast<GlobalOffsetT>(num_items);
+         current_partition_offset += static_cast<GlobalOffsetT>(max_partition_size))
+    {
+      const GlobalOffsetT remaining_items = (num_items - current_partition_offset);
+      const GlobalOffsetT current_num_items =
+        (remaining_items < max_partition_size) ? remaining_items : max_partition_size;
+
+      d_indexed_offset_in = arg_index_input_iterator_t(
+        detail::reduce::make_offset_iterator(d_in, constant_offset_it_t{current_partition_offset}));
+
+      error = dispatch_reduce_t::Dispatch(
```

Comment:
Important: if any of the segment reduction fails, the `error` is not propagated here.

Reply:
Great catch!